### PR TITLE
[ADR] New Authorization roles

### PR DESCRIPTION
**Pull request by enrichman**
See: https://github.com/epinio/epinio/issues/2631

- Docs: https://github.com/epinio/docs/pull/360
- Epinio: https://github.com/epinio/epinio/pull/2656
- Charts: https://github.com/epinio/helm-charts/pull/506

### CODE REVIEW

File: docs/decisions/20231013-authorization-roles.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+## Considered Options
+
+- Secret with whitelist ACL per namespace
+- Role Bindings as their own CRDs
+- ConfigMaps defined Roles with predefined Actions
+
+## Decision Outcome
+
+Chosen option 3 (__"ConfigMaps defined Roles with predefined Actions"__). It was the easier and more flexible solution. It will not require many changes in the code, and it will be easy to customize and implement.
```
Comment (andreas-kupries) [inline docs/decisions/20231013-authorization-roles.md:34]:
```suggestion
Chosen option 3 (__"ConfigMaps defined Roles with predefined Actions"__). It is believed to be the easiest and most flexible solution. It will not require many changes in the code, and it will be easy to customize and implement.
```

File: docs/decisions/20231013-authorization-roles.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+Additionally, as long as the secret/configmap is mounted in the pod without any subpath, changing it would not require a restart of the API.
+
+#### Pros
+
+* Clear "resource" and "permission" description
+
+#### Cons
+
+* Not sure how to handle newly created namespace
```
Comment (andreas-kupries) [inline docs/decisions/20231013-authorization-roles.md:76]:
```suggestion
* Not sure how to handle newly created namespaces
```

File: docs/decisions/20231013-authorization-roles.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+* Some operations are not strictly tight to a "resource". How to avoid the `exec` but allowing the creation of apps?
```
Comment (andreas-kupries) [inline docs/decisions/20231013-authorization-roles.md:77]:
```suggestion
* Some operations are not strictly tied to a "resource". How to avoid the `exec` but allowing the creation of apps?
```

File: docs/decisions/20231013-authorization-roles.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+* Clear and structured implementation
+
+#### Cons
+
+* Difficult/complex to implement and evolve
+* Complex to manage
+
+### ConfigMaps defined Roles with predefined Actions
+
+To avoid changing too much the current implementation we can add "roles" to an Epinio user. At the moment the user can only have one role, but this can be extended with a list instead of a fixed value. Also the role can be namescoped adding a `::` delimiter.
```
Comment (andreas-kupries) [inline docs/decisions/20231013-authorization-roles.md:94]:
```suggestion
To avoid changing the current implementation too much we can add "roles" to an Epinio user. While a user can only have one role at the moment, this can be extended in the future with a list instead of a single value. Furthermore a role can be namespace-scoped by adding a `::` delimiter.
```

Comment (andreas-kupries) [reply docs/decisions/20231013-authorization-roles.md:94]: Question here. Why `::` for the delimiter ? What other delimiters were considered ? (`/`, `:`, `,`, ...).

File: docs/decisions/20231013-authorization-roles.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+```yaml
+metadata:
+  annotations:
+    epinio.io/roles: epinio-role-reader,admin::workspace
+  labels:
+    epinio.io/role: user
+```
+
+*Note:*
+an annotation will be used because labels have some limitations about the lenght and we don't want to be restricted about the number of roles. Also we don't need to perform any kind of lookup over the role.
```
Comment (andreas-kupries) [inline docs/decisions/20231013-authorization-roles.md:107]:
```suggestion
An annotation will be used because labels have length-limitations and we don't want to be restricted about the number of roles. Also we don't need to perform any kind of lookup over the role.
```

File: docs/decisions/20231013-authorization-roles.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+In the previous example the user will have the `epinio-role-reader` and `admin::workspace` roles. When working in the workspace namespace the `admin` role will be used, otherwise the `epinio-role-reader` will take over.
```
Comment (andreas-kupries) [inline docs/decisions/20231013-authorization-roles.md:109]:
```suggestion
In the previous example the user has the roles `epinio-role-reader` and `admin::workspace`. When working in the  namespace `workspace` the `admin` role will be used, otherwise the role `epinio-role-reader`.
```

File: docs/decisions/20231013-authorization-roles.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+A Role can be simply defined as a ConfigMap (no need to be a secret) with a special label (`epinio.io/role: "true"`). This role contains the `actions` that this role can perform, and some metadata that can be used as descriptions (i.e.: the `name`):
```
Comment (andreas-kupries) [inline docs/decisions/20231013-authorization-roles.md:111]:
```suggestion
A `Role` is defined as a kubernetes `ConfigMap` (no need to be a `Secret`) with a special label (`epinio.io/role: "true"`). This resource contains the `actions` that the role can perform, and some metadata that can be used for descriptions (i.e.: the `name`):
```

File: docs/decisions/20231013-authorization-roles.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    epinio.io/role: "true"
+  name: epinio-role-namespace
+  namespace: epinio
+data:
+  name: Epinio Namespace Role
+  actions: |
+    namespace
+```
+
+This actions could be **hardcoded** in a yaml file. This simplify the management of them, and also the flexibility. An action can have some "dependencies", i.e.: the `namespace` action is a union of the `namespace_show`, `namespace_delete`, and so on.
```
Comment (andreas-kupries) [inline docs/decisions/20231013-authorization-roles.md:127]:
```suggestion
The actions are **hardcoded** as an embedded yaml file. This simplifies their management, and also enhances flexibility. An action can have some "dependencies", i.e.: the `namespace` action is a union of the `namespace_show`, `namespace_delete`, and so on.
```

File: docs/decisions/20231013-authorization-roles.md [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+Every action will have a set of endpoints that will allow. Some Epinio operations are formed by multiple endpoints, i.e. the `app_push` consists of a Create, Update and others.
```
Comment (andreas-kupries) [inline docs/decisions/20231013-authorization-roles.md:129]:
```suggestion
Every action lists the set of endpoints it allows. Note that some Epinio operations are formed from multiple endpoints, i.e. the `app_push` consists of a Create, Update and others.
```

File: docs/decisions/20231013-authorization-roles.md [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+* Clear and structured implementation
+
+#### Cons
+
+* Difficult/complex to implement and evolve
+* Complex to manage
+
+### ConfigMaps defined Roles with predefined Actions
+
```
Comment (andreas-kupries) [inline docs/decisions/20231013-authorization-roles.md:93]: Generally speaking this part needs quite a bit more elaboration of syntax and semantics of the config maps, the action.yaml, the actions (if there is a difference).

Treat this part like writing a specification with __all__ the details.
Currently a lot seems to be missing and glossed over.

File: docs/decisions/20231013-authorization-roles.md (hunk 4 above)
Comment (enrichman) [reply docs/decisions/20231013-authorization-roles.md:94]: I was looking for a character that was not "reserved", or conflicting with the Kubernetes rules (btw annotations are more flexible I think). Commas are usually used to separate completely different "set", and it was already used to separate different roles. I was looking for something that seemed to be like a URN.

While testing the OIDC authentication with Github the returned groups are formed with the `org:team` pair, so it looked ok to have a similar (but different) delimiter for `role::namespace`. Also `/` could have been fine btw, but the roles annotations would have looked "weird". Checking the yaml basically the choices would have been:

```yaml
metadata:
  annotations:
    epinio.io/roles: "admin,user::workspace,reader::foobar"
---
metadata:
  annotations:
    epinio.io/roles: "admin,user/workspace,reader/foobar"
---
metadata:
  annotations:
    epinio.io/roles: "admin,user:workspace,reader:foobar"
```

This is the outcome of the `rolesMapping` section in the dex.yaml configuration:

```yaml
    - connectorId: github
      groups:
        - id: epinio:epinio
          roles:
            - admin
        - id: epinio:ui
          roles:
            - user::workspace
        - id: Org1:TeamBlue
          roles:
            - user
            - admin::workspace
```

I'm ok with changing it to something different. Maybe the single `:` (or triple? `:::`) , or the `/` also looks nice to me.

Regarding delimiters I was not sure also about the comma. Maybe a pipe `|` could have been an option easier to read:

```yaml
metadata:
  annotations:
    epinio.io/roles: "admin|user::workspace|reader::foobar"
---
metadata:
  annotations:
    epinio.io/roles: "admin|user:workspace|reader:foobar"
---
metadata:
  annotations:
    epinio.io/roles: "admin|user/workspace|reader/foobar"
```

Comment (andreas-kupries) [reply docs/decisions/20231013-authorization-roles.md:94]: I am ok with the comma for role separation. I would have used `:` or `/` for the namespace separation.

Comment (enrichman) [reply docs/decisions/20231013-authorization-roles.md:94]: Ok, I think I will change it to a single `:` then. :+1:

File: docs/decisions/20231013-authorization-roles.md [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+| Action ID         | Description 
+|-------------------|-------------
+| `gitconfig_read`    | Read permissions (list, show)
+| `gitconfig_write`   | Write permissions (create, delete)<br/>Depends on: `gitconfig_read`
+| `gitconfig`         | All the above<br/>Depends on: `gitconfig_read`, `gitconfig_write`
+
+##### Export Registries
+
+This action enable operations on Export Registries commands and resources. Only read operations are available.
```
Comment (andreas-kupries) [inline docs/decisions/20231013-authorization-roles.md:205]:
```suggestion
This action enables operations on Export Registries commands and resources. Only read operations are available.
```

**Review by andreas-kupries — CHANGES_REQUESTED** [review]
A nit.

**Comment by mmartin24** [comment]
Base tests are covered here: https://github.com/epinio/epinio/issues/2678 and it looks okay to me including some edge cases initially spotted. Please bear in mind this is the first iteration of tests. Given the magnitude of this feature, more edge cases may appear and are subject to be added in the future.
